### Task 1: Permutations and Combinations
Suppose we alter the Lady Tasting Tea experiment to involve twelve cups of tea. Six have the milk in first and the other six having tea in first. A person claims they have the special power of being able to tell whether the tea or the milk went into a cup first upon tasting it. You agree to accept their claim if they can tell which of the six cups in your experiment had the milk in first.

Calculate, using Python, the probability that they select the correct six cups. Here you should assume that they have no special powers in figuring it out, that they are just guessing. Remember to show and justify your workings in code and MarkDown cells.

Suppose, now, you are willing to accept one error. Once they select the six cups they think had the milk in first, you will give them the benefit of the doubt should they have selected at least five of the correct cups. Calculate the probability, assuming they have no special powers, that the person makes at most one error.

Would you accept two errors? Explain.

In [129]:
# Import dependencies ------
# Mathematical functions from the standard library.
# https://docs.python.org/3/library/math.html
import math
# Permutations and combinations.
# https://docs.python.org/3/library/itertools.html
import itertools
# Random selections.
# https://docs.python.org/3/library/random.html
import random
# Numerical structures and operations.
# https://numpy.org/doc/stable/reference/index.html#reference
import numpy as np
# Plotting.
# https://matplotlib.org/stable/contents.html
import matplotlib.pyplot as plt

# set variables ----------

# Number of cups of tea in total.
no_cups = 12 #k
# Number of cups of tea with milk in first.
no_cups_milk_first = 6 #n
# Number of cups of tea with tea in first.
no_cups_tea_first = 6 #n-k

The total number of possible combinations is calculated with combination formula (binomial coefficient )or the number of ways to choose 𝑘 items from a set of 𝑛 items, where order does not matter:
>$\binom{n}{k} = \frac{n!}{k!(n-k)!}$

- 𝑛! is the factorial of 𝑛, the total number of items.
- 𝑘! is the factorial of 𝑘, the number of items you're choosing.
- (𝑛−𝑘)! is the factorial of the remaining items.


In [130]:
# evaluation the choose formula (ref: lecture notes)
n,k = no_cups, no_cups_milk_first #assigning variables
choose = math.factorial(n)/(math.factorial(k)*(math.factorial(n-k)))
choose


924.0

In [131]:
# math.comb function does the same thing. (ref: lecture notes)
ways = math.comb(no_cups, no_cups_milk_first)
# Show.
ways

924

In [132]:
# we can show all possible combinations the 6 cups can be chosen
    # with list(itertools.combinations()function) 
# (ref: lecture notes)

# Create The cup labels.
labels = list(range(n)) #create a list containing integers from 0 to 7 to represent the cups
print("teacup labels: ", labels)
# Show the different ways (combinations) of selecting no_cups_milk_first out of no_cups cups of tea.
combinations = list(itertools.combinations(labels, k))
# Show.
combs=combinations
print("count combinations: ",len(combs))
combs

teacup labels:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
count combinations:  924


[(0, 1, 2, 3, 4, 5),
 (0, 1, 2, 3, 4, 6),
 (0, 1, 2, 3, 4, 7),
 (0, 1, 2, 3, 4, 8),
 (0, 1, 2, 3, 4, 9),
 (0, 1, 2, 3, 4, 10),
 (0, 1, 2, 3, 4, 11),
 (0, 1, 2, 3, 5, 6),
 (0, 1, 2, 3, 5, 7),
 (0, 1, 2, 3, 5, 8),
 (0, 1, 2, 3, 5, 9),
 (0, 1, 2, 3, 5, 10),
 (0, 1, 2, 3, 5, 11),
 (0, 1, 2, 3, 6, 7),
 (0, 1, 2, 3, 6, 8),
 (0, 1, 2, 3, 6, 9),
 (0, 1, 2, 3, 6, 10),
 (0, 1, 2, 3, 6, 11),
 (0, 1, 2, 3, 7, 8),
 (0, 1, 2, 3, 7, 9),
 (0, 1, 2, 3, 7, 10),
 (0, 1, 2, 3, 7, 11),
 (0, 1, 2, 3, 8, 9),
 (0, 1, 2, 3, 8, 10),
 (0, 1, 2, 3, 8, 11),
 (0, 1, 2, 3, 9, 10),
 (0, 1, 2, 3, 9, 11),
 (0, 1, 2, 3, 10, 11),
 (0, 1, 2, 4, 5, 6),
 (0, 1, 2, 4, 5, 7),
 (0, 1, 2, 4, 5, 8),
 (0, 1, 2, 4, 5, 9),
 (0, 1, 2, 4, 5, 10),
 (0, 1, 2, 4, 5, 11),
 (0, 1, 2, 4, 6, 7),
 (0, 1, 2, 4, 6, 8),
 (0, 1, 2, 4, 6, 9),
 (0, 1, 2, 4, 6, 10),
 (0, 1, 2, 4, 6, 11),
 (0, 1, 2, 4, 7, 8),
 (0, 1, 2, 4, 7, 9),
 (0, 1, 2, 4, 7, 10),
 (0, 1, 2, 4, 7, 11),
 (0, 1, 2, 4, 8, 9),
 (0, 1, 2, 4, 8, 10),
 (0, 1, 2, 4, 8, 11),
 (0, 1, 2, 4

Now we have 3 different ways of proving that there are 924 possible combinations of chosing 6 items out of 12.
$$\binom{n}{k} = \frac{n!}{k!(n-k)!} $$
$$\binom{12}{6} = \frac{12!}{6!(12-6)!} = 924$$


##### *Probability*
In this case, choosing the 6 correct cups from 12 is a mutually exclusive scenario where the lady either gets it all right or not. There are 924 different ways to choose 6 cups from 12 and only 1 way to get it right when all 6 selected cups are milk-in-first cups. Therefore, she has a 1 in 924 (or 0.11%) chance of succeeding. 
$$ P= \frac {1}{\binom {n}{k}} = \frac {1}{\binom {12}{6}} = \frac {1}{924} $$

###### ref: combination formula: https://en.wikipedia.org/wiki/Combination<br>ref: binomial coefficient: https://en.wikipedia.org/wiki/Binomial_coefficient <br> ref: calculate probability, mutually exclusive: https://www.wikihow.com/Calculate-Probability

In [133]:
# Percentage chance
P = round((1/choose)*100,2)
print(f"{P}%")

0.11%


The lady has a very low 0.11% chance to select the correct cups.

Now let's see what is the probability of success if we allow 1 error so we accept both 5 correct and 6 correct cups as successful trial.

##### Using the combination formula $\binom{n}{k} = \frac{n!}{k!(n-k)!} $
- Total number of combinations (6 cups of 12 cups in total): $\binom{12}{6} = \frac{12!}{6!(12-6)!} =924$ <br>
- Total number of correct answers (6 cups of 6 milk-in-first cups): $\binom{6}{6} = \frac{6!}{6!(6-6)!} =1$ <br>

- Total number of answers with 5 correct (5 correct of 6 correct): $\binom{6}{5} = \frac{6!}{5!(6-5)!} =6$ <br>
- Total number of answers with 1 wrong (1 wrong of 6 wrong): $\binom{6}{1} = \frac{6!}{1!(6-1)!} =6$ <br>
>Since every time the lady selects an incorrect cup, there are 6 ways to choose 5 correct cups, and there are 6 possible incorrect cups, the total number of combinations for selecting 1 incorrect cup along with 5 correct cups is $6×6=36$ and therefore the total number of combinations to pick the correct cups allowing for 1 error is $36+1=37$. ($+1$ being the number of possible combinaions for all correct cups selected.)

In [148]:
# create function to apply combination formula C=n!/k!x(n-k)! 
# this is doing the same as math.comb(n,k) but with visible anatomy.
def combinations(n,k):
    C= math.factorial(n)//(math.factorial(k)*(math.factorial(n-k)))
    return(C)

# Allowing for 1 error

# number of possible combinationsinations using the combinationsination formula
tot=combinations(12,6)
# number of possible combinationsinations for correct answer
correct=combinations(6,6) # this is 1
# number of possible combinationsinations with 1 error (5 correct of 6 out of 12)
correct5=combinations(6,5) # this is 6
# number of possible combinationsinations with 1 error (1 incorrect of 6 out of 12)
error1=combinations(6,1) # also 6

print("one error:",error1,"\nfive correct:", correct5)
print(f"P(robability) = (AllCorrect + Correct5 * Error1) / Total")
print(f"P = ({correct} + {correct5} * {error1}) / {tot}")
print(f"P ={correct+(correct5*error1)}/{tot} = {round((correct+(correct5*error1))/tot*100,2)}%")



one error: 6 
five correct: 6
P(robability) = (AllCorrect + Correct5 * Error1) / Total
P = (1 + 6 * 6) / 924
P =37/924 = 4.0%


The lady has a 4% chance of choosing 5 or 6 correct cups with milk-in-first out of 12 cups which is still quite a low probability.

In [146]:

# Allowing for 2 errors

# number of possible combinationsinations with 1 error (4 correct of 6 out of 12)
correct4=combinations(6,4)
# number of possible combinationsinations with 1 error (2 incorrect of 6 out of 12)
error2=combinations(6,2)

#number of one-error combinations 37
# number of 2 error combination is adding the combinations of 2 errors and 4 corrects to the 
# previously calcualted no error and one error combinations 1(no error)+36(1 error)+226(2 errors (15x15)=262

print("two errors:",error2,"\nfour correct:", correct4)

print(f"P = (AllCorrect + Correct5 * Error1 + Error2 * Correct4) / Total")
print(f"P = ({correct} + {correct5} * {error1} + {error2} * {correct4}) / {tot}")
print(f"P={correct+(correct5*error1)+(error2*correct4)}/{tot} = {round((correct+(correct5*error1)+(error2*correct4))/tot*100,2)}%")


two errors: 15 
four correct: 15
P = (AllCorrect + Correct5 * Error1 + Error2 * Correct4) / Total
P = (1 + 6 * 6 + 15 * 15) / 924
P=262/924 = 28.35%


In conclusion, the lady has over 28% chance to select the cups with 2 errors allowed, 4% to select with 1 error and 0.11% to select all 6 cups correctly out of 12.
<br>
- *Null hypothesis (H₀):* The lady has no special power to tell the difference between cups of tea with milk in first and cups of tea with tea in first. The lady is selecting the six cups from the twelve cups at random.
- *Alternative hypothesis (H₁):* The lady has a supernatural "cup-with-milk-in-first" sense.

### Conclusion

Your paragraph is clear and concise! Here’s a slightly polished version with minor adjustments for clarity and flow:

If the lady is just guessing, it’s unlikely she will pick the correct six cups with milk first, with a chance of only 0.11%. This low probability suggests that a successful selection of the six cups (especially multiple times) indicates something non-random is occurring, supporting the alternative hypothesis. Allowing for one error increases the probability to 4%, which is still a very low likelihood and below the conventional 5% threshold. (This $𝑝=0.05$ level of probability was first introduced by Sir Ronald Fisher in his 1925 book, Statistical Methods for Research Workers (1) for establishing statistical significance.)
<br>
However, allowing for a second error raises the probability to over 28%, which exceeds the conventional threshold. With this, the lady has a high chance of about 1 in 4 of getting it right, making the experiment unsuitable for rejecting the null hypothesis as it is quite likely she may randomly select at least 4 correct cups.


###### ref: Cowles, M. and Davis, C. (1982). On the origins of the .05 level of statistical significance. American Psychologist, 37(5), pp.553–558. doi:https://doi.org/10.1037//0003-066x.37.5.553.

# Task 2: numpy's Normal Distribution
In this task you will assess whether numpy.random.normal() properly generates normal values. 
>To begin, generate a sample of one hundred thousand values using the function with mean 10.0 and standard deviation 3.0.
<br>
Use the scipy.stats.shapiro() function to test whether your sample came from a normal distribution. Explain the results and output.
<br>
Plot a histogram of your values and plot the corresponding normal distribution probability density function on top of it.
